### PR TITLE
Refactor ConfigurationUtils

### PR DESCRIPTION
**Pull request by rongrong**
* Make reloadProperties more efficient
* Rename ConfigurationUtils.defaults to copyDefaults so copy is explicit
* Add no-copy implementation of ConfigurationUtils.defaults

### CODE REVIEW

File: core/common/src/main/java/alluxio/conf/InstancedConfiguration.java [hunk 1]
```diff
@@ -70,7 +70,7 @@ public class InstancedConfiguration implements AlluxioConfiguration {
    * @return an instanced configuration preset with defaults
    */
   public static InstancedConfiguration defaults() {
-    return new InstancedConfiguration(ConfigurationUtils.defaults());
+    return new InstancedConfiguration(ConfigurationUtils.copyDefaults());
```
Comment (LuQQiu) [inline core/common/src/main/java/alluxio/conf/InstancedConfiguration.java:73]: copyDefaults here?

File: stress/shell/src/main/java/alluxio/stress/cli/suite/MasterMaxThroughput.java [hunk 2]
```diff
@@ -77,7 +77,7 @@ public void prepare() throws Exception {
     }
     try (JobMasterClient client = JobMasterClient.Factory.create(
         JobMasterClientContext.newBuilder(ClientContext.create(new InstancedConfiguration(
-            ConfigurationUtils.defaults()))).build())) {
+            ConfigurationUtils.copyDefaults()))).build())) {
```
Comment (LuQQiu) [inline stress/shell/src/main/java/alluxio/stress/cli/suite/MasterMaxThroughput.java:80]: why not defaults?

File: table/shell/src/test/java/alluxio/cli/table/TransformTableCommandTest.java [hunk 3]
```diff
@@ -41,7 +41,7 @@ private void transformInternal(String definition, String expected) throws Except
     when(client.transformTable(ArgumentMatchers.anyString(), ArgumentMatchers.anyString(),
         ArgumentMatchers.anyString())).thenReturn(0L);
     TransformTableCommand command =
-        new TransformTableCommand(new InstancedConfiguration(ConfigurationUtils.defaults()),
+        new TransformTableCommand(new InstancedConfiguration(ConfigurationUtils.copyDefaults()),
```
Comment (LuQQiu) [inline table/shell/src/test/java/alluxio/cli/table/TransformTableCommandTest.java:44]: why not defaults?

File: underfs/local/src/test/java/alluxio/underfs/local/LocalUnderFileSystemTest.java [hunk 4]
```diff
@@ -60,7 +60,7 @@ public class LocalUnderFileSystemTest {
   private String mLocalUfsRoot;
   private UnderFileSystem mLocalUfs;
   private static AlluxioConfiguration sConf =
-      new InstancedConfiguration(ConfigurationUtils.defaults());
+      new InstancedConfiguration(ConfigurationUtils.copyDefaults());
```
Comment (LuQQiu) [inline underfs/local/src/test/java/alluxio/underfs/local/LocalUnderFileSystemTest.java:63]: can use defaults?

File: underfs/obs/src/test/java/alluxio/underfs/obs/OBSInputStreamTest.java [hunk 5]
```diff
@@ -46,7 +46,7 @@ public class OBSInputStreamTest {
   private static final String BUCKET_NAME = "testBucket";
   private static final String OBJECT_KEY = "testObjectKey";
   private static AlluxioConfiguration sConf =
-      new InstancedConfiguration(ConfigurationUtils.defaults());
+      new InstancedConfiguration(ConfigurationUtils.copyDefaults());
```
Comment (LuQQiu) [inline underfs/obs/src/test/java/alluxio/underfs/obs/OBSInputStreamTest.java:49]: defaults?

**Review by LuQQiu — COMMENTED** [review]
Didn't check all the copyDefaults, but in the future PR can verify whether the places with copyDefaults are needed
Thanks for reducing the configuration copies!!!

File: core/common/src/main/java/alluxio/conf/InstancedConfiguration.java (hunk 1 above)
Comment (rongrong) [reply core/common/src/main/java/alluxio/conf/InstancedConfiguration.java:73]: No, `InstancedConfiguration` has `set` method, so it is not safe to just change this to return `defaults()`.

File: stress/shell/src/main/java/alluxio/stress/cli/suite/MasterMaxThroughput.java (hunk 2 above)
Comment (rongrong) [reply stress/shell/src/main/java/alluxio/stress/cli/suite/MasterMaxThroughput.java:80]: These are just from renaming `defaults` to `copyDefaults` I haven't checked through all instances to sort out whether a static copy is good enough vs it requires a copy yet. More cleanup will come later.

File: core/common/src/main/java/alluxio/util/ConfigurationUtils.java [hunk 6]
```diff
@@ -389,22 +333,23 @@ public static String valueAsString(String value) {
...
+  }
+
+  /**
+   * Returns a reference to {@link AlluxioProperties} with the defaults and values from
+   * alluxio-site properties.
+   *
+   * @return a reference to Alluxio properties loaded from the site-properties file
+   */
+  public static AlluxioConfiguration defaults() {
+    return new InstancedConfiguration(DEFAULT_PROPERTIES_REFERENCE.get());
```
[the remark this answers is not in the document]
Comment (LuQQiu) [reply core/common/src/main/java/alluxio/util/ConfigurationUtils.java:352]: Got it
AlluxioConfiguration -> immutable
AlluxioProperties -> updatable
Sounds good

**Comment by rongrong** [comment]
alluxio-bot, merge this please

**Pull request merged**